### DISCUSSION TITLE
Help with QueryCleanup

### DISCUSSION THREAD

**Discussion by lazyclaw-labs** [linq2db/linq2db#4747] [General]
My scenario:
In ValueConverter, I have borrowed some memory frmo Memorypool.
I need to return those memory on Query Completion.

I couldn't find a place that is always run when a Query is done.
I currently use OnCommandDispose to do the clean up job, but it's the wrong place, because if one of the ValueConverter throw an exception, the command won't even be created.

Any suggestions?

**Comment by MaceWindu** [comment]
I don't think there is anything like that. Your best option is to handle following events:

- `ExecuteNonQuery[Async]`
- `ExecuteScalar[Async]`
- `BeforeReaderDispose[Async]`

for all 3 types of queries using [this interceptor](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Interceptors/CommandInterceptor.cs)

If you will have other cases uncovered by those - we can look into fixing it.

**Reply by MaceWindu** [reply]
as for errors, generated before command created - I don't think we have anything in place now. New interceptor event suggestion with test-case will help to move it further

**Comment by lazyclaw-labs** [comment]
I think I've found the solution. There is a IExecuteScope GetScope() on DataProvider. I provide my own ExecuteScope, and cleanup here in its dispose function..
It seems ExecuteScope is invented to specifically solve my problem. :-) 

Do I understand it correctly?

**Reply by MaceWindu** [reply]
Technically it was added to workaround issues with providers, that use current culture instead of invariant culture when converting parameters to strings, but if it works for you - you can use it too

**Comment by lazyclaw-labs** [comment]
Seems like it worked for me. Just want to confirm: it was retrieved at the beginning of a query, and got disposed after the query is done, if the scope is not null of course.


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by lazyclaw-labs** [comment]
Just to be sure, it is always retrieved and always got disposed! Not depending on any other  conditions

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by MaceWindu** [reply]
Yes

**Comment by lazyclaw-labs** [comment]
Nice!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
